src/ad/AdjointsDI/adjoints.jl

```diff
@@ -0,0 +1,250 @@
+            f = forces
+        end
+        if ismissing(p)
+            p = Jutul.setup_parameters(model)
+        end
+        dt = [step_info[:dt]]
+        case = JutulCase(model, dt, f, parameters = p, state0 = state0)
+    end
+    return case
+end
+
+function unpack_setup(step_info, N, case::JutulCase)
+    # Either case for all steps or case for current step
+    Ns = length(case.dt)
+    if Ns > 1
+        if case.forces isa Vector
+            Ns == N || error("case.forces was a vector and expected $N steps, got $Ns.")
+        end
+        case = case[step_info[:step]]
+    end
+    return case
+end
+
+function unpack_setup(step_info, N, out::Tuple)
+    unpack_setup(step_info, N, out...)
+end
+
+function unpack_setup(step_info, N, model::Jutul.JutulModel)
+    return (model, missing, missing, missing)
+end
+
+function unpack_setup(step_info, N, model::Jutul.JutulModel, parameters)
+    return (model, parameters, missing, missing)
+end
+
+function unpack_setup(step_info, N, model::Jutul.JutulModel, parameters, forces)
+    return (model, parameters, forces, missing)
+end
+
+function unpack_setup(step_info, N, model::Jutul.JutulModel, parameters, forces, state0)
+    return (model, parameters, forces, state0)
+end
+
+function setup_jacobian_evaluation!(storage, X, F, G, states, case0, forces, timesteps, backend, do_prep, single_step_sparsity)
+    N = length(timesteps)
+    if ismissing(backend)
+        backend = AutoSparse(
+            AutoForwardDiff();
+            sparsity_detector = TracerLocalSparsityDetector(),
+            coloring_algorithm = GreedyColoringAlgorithm(),
+        )
+    end
+    cache = Dict()
+    function evaluate_for_states(x, state, state0, step_info, dt)
+        # case = F(x, step_info)
+        case = setup_case(x, F, step_info, state0, forces, N)
+        if step_info[:step] == 1
+            state0 = case.state0
+        end
+        sim = HelperSimulator(case, eltype(x), cache = cache, n_extra = 1)
+        model_residual(state, state0, sim,
+            forces = case.forces,
+            time = step_info[:time],
+            dt = dt
+        )
+        r = sim.storage.r_extended
+        r[end] = G(case.model, state, dt, step_info, case.forces)
+        return r
+    end
+    dt = case0.dt[1]
+    function evaluate0(x)
+        info = Jutul.optimization_step_info(1, 0.0, dt, total_time = sum(timesteps))
+        v = evaluate_for_states(x, states[1], case0.state0, info, dt)
+        if !single_step_sparsity
+            for i in 2:N
+                step_info = Jutul.optimization_step_info(i, sum(timesteps[1:(i-1)]), timesteps[i], total_time = sum(timesteps))
+                v .+= evaluate_for_states(x, states[i], states[i-1], step_info, timesteps[i])
+            end
+        end
+        return v
+    end
+    storage[:function_di] = evaluate_for_states
+    # Note: strict = false is needed because we create another function on the fly
+    # that essentially calls the same function.
```

Comment:
Why not create that function only once?

Reply:
The sparsity pattern stays fixed but I have to pass some other values in (acting as parameters). I suspect there is a better way to do this in the API...